enhancements/monitoring/alert-overrides.md

```diff
@@ -0,0 +1,280 @@
+asked for a supported method of adding additional custom rules, and adjusting
+thresholds, or adding labels and annotations to existing platform alerting
+rules.  These modifications are currently not possible, because the alerts are
+managed by operators and cannot be modified, and the addition of custom rules is
+explicitly unsupported at the moment.
+
+## Motivation
+
+OpenShift is an opinionated platform, and the default alerting rules have of
+course been carefully crafted to match what we think the majority of users need.
+Nevertheless, users have repeatedly asked for the ability to modify alerting
+rules to adjust thresholds, or to simply add labels and annotations to aide in
+routing alert notifications, as well as a supported method of adding additional
+alerting rules.
+
+Users currently have the ability to add additional rules from a technical
+standpoint -- by simply creating new `PrometheusRule` objects -- but this is
+explicitly unsupported at the moment.  The reason being that these resources are
+handled directly by the prometheus-operator, making it difficult to perform
+OpenShift specific defaulting and validation.  A concrete example being that
+there is no way to prevent users from creating new recording rules that may
+overwhelm the platform Prometheus instance.
+
+We've also encountered bugs and unexpected interactions that cause alerts to
+fire erroneously.  A supported method of patching platform alerting rules
+provides a more flexible method of applying temporary fixes in such cases.
+
+The goal of this proposal is to enable the addition of custom alerting rules,
+and limited ability to override existing platform alerting rules via a resource
+under the direct control of the cluster-monitoring-operator.
+
+### Goals
+
+- Give users a single, supported, and documented place to add additional rules,
+  in a way that allows the cluster-monitoring-operator to provide its own
+  defaulting, validation, etc.
+- Allow users to apply overrides to alerting rules shipped by the platform.
+- Allow users to drop unwanted platform alerts.
+- User-supplied overrides are merged with platform rules -- updates to fields
+  not modified by users continue to get updates.
+- Resources shipped by operators remain unmodified.  No action needed from
+  OpenShift developers.
+
+### Non-Goals
+
+- Only alerting rules are available for modification.  Not recording rules or
+  other monitoring configuration.
+
+## Proposal
+
+The cluster-monitoring-operator will be extended with a new controller that
+watches a user-supplied `ConfigMap` with a single key having the following
+structure:
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: alert-overrides
+  namespace: openshift-monitoring
+data:
+  overrides.yaml: |
+    overrides: # List of platform rule overrides.
+    - key:
+        name: "Watchdog" # Alerts are identified by name and severity tuple.
+        severity: "none"
```

Comment:
The `name` & `severity` together are not enough to uniquely identify rules (for example 2 rules in different rule groups with the same name and severity).

In order to uniquely identify rules in the Console UI, we use all of: (group file, group name, rule name, rule duration, rule query, rule labels). I think we will need to do the same or something similar here.

Reply:
To extend on this a bit: I think the enhancement proposal could benefit from explicitly stating whether overrides must uniquely identify a single alert or not. Overriding or dropping multiple alerts with one spec could certainly be useful.
However this adds significant complexity from the UI perspective. How can a user ensure their override does what they intend.
If we want an explicit one to one matching lets call that out and sketch how a user interaction would look like if they provide an insufficient spec (i.e. one that doesn't uniquely identify one alert).

Reply:
Yeah, this is a point that absolutely needs clarification, thanks.

On a fresh 4.10-nightly cluster on AWS, I only see two alerts with this problem. One is `AlertmanagerClusterFailedToSendAlerts` which is due to cluster-monitoring-operator explicitly bringing the `critical` version of the alert down to `warning` and the other is `KubeAPIErrorBudgetBurn` which is a bit more complicated. That one legitimately seems to have two versions each at the `warning` and `critical` severities. Of course there could be other instances shipped by add-ons.

I suppose this is a trade-off between preventing an override matching multiple rules, and requiring the user to specify a bunch of fields to match against -- ones that often aren't the most visible to users. Given the above, I don't think either of these options are unreasonable:

1) Consider multiple matches of name + severity as an error condition. Can't override that, sorry. We could couple this with amending the style-guide to disallow this, adding e2e tests for it in `origin`, and fixing the offending alerts.

2) Just apply the override to all matching alerts. Though, like you said, that may present UX challenges.

I guess I would lean towards the first option personally.

Reply:
Having a clear an concise failure mode definitely beats a murky UI, so +1 for 1. Though it might be valuable to at least mention both approaches in this document?

Reply:
another example here is Kube* rules that are generated with a `namespace=` field. SRE use this currently to determine what should/could be ignored.

Reply:
> another example here is Kube* rules that are generated with a namespace= field. SRE use this currently to determine what should/could be ignored.

Sorry, not sure I follow. Do you mean that you'd like to be able to match on the `namespace` label in addition to name and severity when overriding an alert?

Reply:
I've added some content about this failure mode.

Reply:
This makes sense to me. I've updated the proposal with a CRD structure that allows matching against an arbitrary set of static labels, but requires an exact match.